docs/proposals/20230524-support-nodepool-with-host-network-mode.md

```diff
@@ -0,0 +1,158 @@
+kube-system    coredns-7d89d9b6b8-vhqj7                   4m           16Mi
+kube-system    coredns-7d89d9b6b8-fn9c4                   3m           16Mi
+kube-system    kube-scheduler-master1                     6m           16Mi
+kube-system    metrics-server-66dd897cc4-zl2sq            8m           14Mi
+kube-system    metrics-server-59d8dc4bc-54v59             6m           13Mi
+kube-flannel   kube-flannel-ds-cs9dp                      9m           11Mi
+cert-manager   cert-manager-webhook-65b766b5f8-z9k6c      5m           11Mi
+kube-system    kube-proxy-x2v2r                           3m           10Mi
+default        nginx-pod                                  0m           3Mi
+```
+
+### Goals
+
+Implement a new type of NodePool that enables Pods on the nodes to use the host network while not using services, CoreDNS, and other components.
+
+### Non-Goals/Future Work
+
+Refactor `yurtadm join` and so on.
+
+## Proposal
+
+### Definition
+
+* NodePool with host network mode
+  * A new NodePool type where Pods will only use host networking mode, suitable for lightweight and minimally demanding edge nodes.
+
+### Architecture
+
+Here is a example for NodePool in host network mode:
+
+```
+apiVersion: apps.openyurt.io/v1beta1
+kind: NodePool
+metadata:
+  name: beijing
+spec:
+  type: Cloud
+  hostNetwork: true
+```
+
+The work we need to do is as follows:
+
+* **Step1**: First, before adding a node to the NodePool, you need to modify the kubelet configuration of the node. This part of the work can be done by yurtadm join.
+
+/var/lib/kubelet/kubeadm-flags.env
+
+`—network-plugin=cni`
+
+`—cni-conf-dir=/etc/cni/net.d`
+
+For example:
+
+```
+KUBELET_KUBEADM_ARGS="--network-plugin=cni —cni-conf-dir=/etc/cni/net.d --pod-infra-container-image=registry.cn-hangzhou.aliyuncs.com/google_containers/pause:3.5"
+```
+
+Set the network plugin to loopback.
+
+```
+sudo cat <<EOF > /etc/cni/net.d/0-loopback.conf
+{
+  "cniVersion": "0.3.0",
+  "name": "lo",
+  "type": "loopback"
+}
+EOF
+```
+
+Restart kubelet
+
+```
+systemctl daemon-reload & systemctl restart kubelet
+```
+We should add nodepool.openyurt.io/host-network=true label to node in the yurtadm join when node is joined into the cluster. Because flannel/kube-proxy/coredns daemon pod will be run on the node before the NodePool controller add this label on the node.
+
+* Step2: Design a new field for NodePool called hostNetwork=true/false. True indicates that this is a NodePool of hostNetwork mode type.
+* Step3: Modify the NodePool controller to manage the hostNetwork label of all nodes in the node-network-mode NodePool.
+* Step4: Add a webhook to verify if there is a nodepool named xxx for the pods with the `nodepool.openyurt.io/desired-nodepool: xxx` tag. If it exists, check if the hostNetwork field of the nodepool is true. If it is true, set the pod's hostNetwork to true. Users can add the pod label themselves or use YurtAppSet and YurtAppDaemon to deploy the application.
```

Comment:
we can add a hostNetwork filter in yurthub to deal with changing hostNetwork field for pods that run on the hostNetwork mode node. so we can handle all pods even if pods are not controlled by YurtAppSet and YurtAppDaemon.